Copy the oneAPI Release Package MPI examples to this repository (#4)
codeplaysoftware/SYCL-samples: README.md

by dragging the mouse and using the mouse wheel to control the camera.

### Fluid Simulation
This demo visualizes fluid behavior in a closed container. Each cell in the
cellular automata represents a fluid particle existing in a velocity field.
Drag the mouse around the screen to create fluid particles with velocities in
direction of the mouse travel. The fluid fades slowly over time so as not to fill
the container.

### MPI for CUDA Backend 

> The source code within provides two SYCL example programs; send_recv_buff.cpp, send_recv_usm.cpp, which are CUDA MPI aware. One example uses the SYCL Unified Share Memory (USM). The other uses the Buffer (buff) model. Each example uses the programing pattern Send-Receive. 

Reviewer: ```suggestion
The source code within provides two SYCL example programs; send_recv_buff.cpp, send_recv_usm.cpp, which use CUDA-Aware MPI. One example uses the SYCL Unified Share Memory (USM). The other uses the Buffer (buff) model. Each example uses the programing pattern Send-Receive. 
```

nit: personally I prefer to not expand USM to Unified Shared Memory because it would make me think it is only for malloc_shared: i.e I think avoiding its terrible naming being made quite so obvious might make it a bit less confusing.

Author: Fix reword 'use CUDA-Aware'.
Done.

A third source code example scatter_reduce_gather demonstrates a common HPC programming idiom using Scatter, Reduce and Gather. A data array is scattered by two processes associated with different MPI ranks using Scatter. The initial data is updated within each MPI rank. Next the updated data is used to calculate a local quantity that is then reduced to a partial result in each rank using the SYCL 2020 reduction interface. Finally, the partial results from each rank are reduced to a final scalar value, ```res```, using Reduce. Finally, the initial data is updated using Gather.

> These three examples form part of the [Codeplay oneAPI for NVIDIA GPUs plugin documentation](https://developer.codeplay.com/products/oneapi/nvidia/2023.2.1/guides/MPI-guide).

Reviewer: ```suggestion
These three examples form part of the [Codeplay oneAPI for NVIDIA GPUs plugin documentation](https://developer.codeplay.com/products/oneapi/nvidia/2024.0.0/guides/MPI-guide).
```

Best to use the latest release doc: really there should be a "latest" link, but AFAIK it doesn't exist.

Reviewer: Probably worth adding a line saying: "Refer to this documentation for a gpu-aware MPI guide for the CUDA backend" or similar.

Author: Fixed the URL link to be 2024.0.0. Added a sentence as recommended.
Done.

by dragging the mouse and using the mouse wheel to control the camera.

### Fluid Simulation
This demo visualizes fluid behavior in a closed container. Each cell in the
cellular automata represents a fluid particle existing in a velocity field.
Drag the mouse around the screen to create fluid particles with velocities in
direction of the mouse travel. The fluid fades slowly over time so as not to fill
the container.

### MPI for CUDA Backend 

> The three minimal code examples demonstrate how GPUs which are

Reviewer: ```suggestion
The three minimal code examples demonstrate how GPUs which support
```

Author: Change made.
Done.

the container.

### MPI for CUDA Backend 
The source code within provides two SYCL example programs; send_recv_buff.cpp, 

> send_recv_usm.cpp, which use CUDA-Aware MPI. One example uses the SYCL Unified 

Reviewer: For me this sounds like only send_recv_buff.cpp and send_recv_usm.cpp use CUDA-Aware MPI and makes it sound like maybe scatter_reduce_gather uses standard host MPI pointers.

The main information that the README.md should have in some form is that:
- The intention of the send_recv_usm.cpp and send_recv_buff.cpp are to be minimal examples of how to use gpu-aware MPI in sycl using either usm or buffers. 
- The third example also uses cuda-aware MPI and USM (since this is much more commonly used than buffers) but is slightly more complex.

Author: Expanded the description to be more explanatory.
Done.
